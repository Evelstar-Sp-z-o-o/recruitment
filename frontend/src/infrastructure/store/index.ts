import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
