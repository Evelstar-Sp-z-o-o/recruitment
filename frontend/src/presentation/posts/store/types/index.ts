type PostsResponse = {
  data: Data;
  id: number;
};

type Data = {
  body: string;
  author: string;
  created: number;
  edited: number;
  postId: string;
};

export type { PostsResponse };
