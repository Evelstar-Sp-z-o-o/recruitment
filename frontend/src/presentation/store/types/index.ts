import Order from '../constants/order';

type PostsResponse = {
  data: Data;
  id: number;
};

type Data = {
  body: string;
  author: string;
  created: number;
  edited: number;
  postId: string;
};

type PostsRequest = {
  sort: string;
  order: Order;
  page: number;
};

type AddPostRequest = {
  body: string;
  author: string;
  created: number;
  postId: string;
  edited?: number | null;
};

type DeletePostRequest = {
  id: number;
};

type EditPostRequest = {
  body: string;
  author: string;
  created: number;
  postId: string;
  id: number;
  edited: number;
};

export type { PostsResponse, AddPostRequest, PostsRequest, DeletePostRequest, EditPostRequest };
