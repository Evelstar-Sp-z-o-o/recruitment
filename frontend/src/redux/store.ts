import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
