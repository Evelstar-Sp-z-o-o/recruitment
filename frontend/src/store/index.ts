import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
