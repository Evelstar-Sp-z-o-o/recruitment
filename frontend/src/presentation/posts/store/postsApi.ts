import apiService from '@/src/infrastructure/store/services/apiService';

import { PostsResponse } from './types';

const POSTS = 'posts';

const postsApi = apiService.enhanceEndpoints({ addTagTypes: [POSTS] }).injectEndpoints({
  endpoints(builder) {
    return {
      getPosts: builder.query<PostsResponse[], void>({
        query: () => ({
          url: '/api/posts',
        }),
        providesTags: [POSTS],
      }),
    };
  },
});

export default postsApi;

export const { useGetPostsQuery } = postsApi;
