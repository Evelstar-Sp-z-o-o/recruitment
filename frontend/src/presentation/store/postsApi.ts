import apiService from '@/src/infrastructure/store/services/apiService';

import { ITEMS_PER_PAGE } from './constants/pagination';
import POSTS from './constants/tags';
import { AddPostRequest, DeletePostRequest, EditPostRequest, PostsRequest, PostsResponse } from './types';

const postsApi = apiService.enhanceEndpoints({ addTagTypes: [POSTS] }).injectEndpoints({
  endpoints(builder) {
    return {
      getPosts: builder.query<PostsResponse[], PostsRequest>({
        query: ({ sort, order, page }) => ({
          url: `/api/posts?_sort=${sort}&_order=${order}&_page=${page}&_limit=${ITEMS_PER_PAGE}`,
        }),
        serializeQueryArgs: ({ endpointName }) => {
          return endpointName;
        },
        merge: (currentCache, newItems, { arg: { page } }) => {
          if (page === 1) {
            return newItems;
          }
          return [...currentCache, ...newItems.filter((item) => !currentCache.some((cached) => cached.id === item.id))];
        },
        forceRefetch({ currentArg, previousArg }) {
          return currentArg !== previousArg;
        },
        providesTags: (result) =>
          result
            ? [...result.map(({ id }) => ({ type: POSTS, id })), { type: POSTS, id: 'LIST' }]
            : [{ type: POSTS, id: 'LIST' }],
      }),
      addPost: builder.mutation<PostsResponse, AddPostRequest>({
        query: (data) => ({
          url: '/api/posts',
          method: 'post',
          body: { data },
        }),
        onQueryStarted: async (post, { dispatch, queryFulfilled }) => {
          try {
            const { data: newPost } = await queryFulfilled;
            dispatch(
              postsApi.util.updateQueryData('getPosts', undefined, (draft) => {
                draft.unshift({ id: newPost.id, data: newPost.data });
              }),
            );
          } catch {
            console.error('Failed to add new post');
          }
        },
      }),
      deletePost: builder.mutation<void, DeletePostRequest>({
        query: ({ id }) => ({
          url: `/api/posts/${id}`,
          method: 'delete',
        }),
        async onQueryStarted({ id }, { dispatch, queryFulfilled }) {
          const patchResult = dispatch(
            postsApi.util.updateQueryData('getPosts', undefined, (draft) => {
              if (draft) {
                const index = draft.findIndex((post) => post.id === id);
                if (index !== -1) {
                  draft.splice(index, 1);
                }
              }
            }),
          );
          try {
            await queryFulfilled;
          } catch {
            patchResult.undo();
          }
        },
      }),
      editPost: builder.mutation<PostsResponse, EditPostRequest>({
        query: ({ id, ...data }) => ({
          url: `/api/posts/${id}`,
          method: 'put',
          body: { data: { ...data } },
        }),
        async onQueryStarted({ id, ...patch }, { dispatch, queryFulfilled }) {
          const patchResult = dispatch(
            postsApi.util.updateQueryData('getPosts', undefined, (draft) => {
              const index = draft.findIndex((post) => post.id === id);
              if (index !== -1) {
                draft[index] = {
                  id: id,
                  data: patch,
                };
              }
            }),
          );
          try {
            const { data: updatedPost } = await queryFulfilled;
            dispatch(
              postsApi.util.updateQueryData('getPosts', undefined, (draft) => {
                const index = draft.findIndex((post) => post.id === id);
                if (index !== -1) {
                  draft[index] = updatedPost;
                }
              }),
            );
          } catch {
            patchResult.undo();
          }
        },
      }),
    };
  },
});

export default postsApi;

export const { useGetPostsQuery, useAddPostMutation, useDeletePostMutation, useEditPostMutation } = postsApi;
