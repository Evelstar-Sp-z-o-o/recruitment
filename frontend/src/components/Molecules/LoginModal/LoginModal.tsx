import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';

import CloseModal from '@/src/components/Atoms/CloseModal/CloseModal';
import { setUser } from '@/src/store';
import { Button, TextField } from '@mui/material';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Modal from '@mui/material/Modal';
import Typography from '@mui/material/Typography';

const StyledModal = {
  width: 300,
  bgcolor: 'background.paper',
};

interface ILoginModalProps {
  isOpen: boolean;
  handleClose: () => void;
  handleSnackbar: () => void;
}

const LoginModal: FC<ILoginModalProps> = ({ isOpen, handleClose, handleSnackbar }) => {
  const dispatch = useDispatch();
  const [isEmailCorrect, setIsEmailCorrect] = useState<boolean>(true);
  const { t } = useTranslation();

  const submitLogin = (e) => {
    e.preventDefault();
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const formData = new FormData(e.currentTarget);
    const formJson = Object.fromEntries((formData as any).entries());
    const email: string = formJson.email;

    if (!emailRegex.test(email)) {
      setIsEmailCorrect(false);
      return;
    }

    handleClose(true);
    handleSnackbar(true);
    localStorage.removeItem('wasHomeModalShown');
    sessionStorage.setItem('user', email);

    return dispatch(setUser(email));
  };

  return (
    <Modal open={isOpen}>
      <>
        <Box sx={StyledModal} className="modal" role="dialog">
          <CloseModal handleClose={handleClose} />
          <Typography variant="h6" component="h2">
            {t('login.header')}
          </Typography>
          <Container sx={{ mt: 2 }}>
            <Container
              component="form"
              sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', p: '0 !important' }}
              onSubmit={submitLogin}
            >
              <TextField
                error={!isEmailCorrect}
                required
                inputProps={{
                  type: 'email',
                }}
                helperText={isEmailCorrect ? ' ' : t('login.incorrect')}
                label={t('login.email')}
                name="email"
                variant="standard"
              />{' '}
              <Button type="submit" variant="outlined">
                {t('login.button')}
              </Button>
            </Container>
          </Container>
        </Box>
      </>
    </Modal>
  );
};

export default LoginModal;
