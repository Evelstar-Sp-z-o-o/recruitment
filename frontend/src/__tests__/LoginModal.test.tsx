import { Provider } from 'react-redux';

import LoginModal from '@/src/components/Molecules/LoginModal/LoginModal';
import { store } from '@/src/store/index';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

describe('LoginModal Component', () => {
  const handleClose = vi.fn();
  const handleSnackbar = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly when open', () => {
    render(
      <Provider store={store}>
        <LoginModal isOpen={true} handleClose={handleClose} handleSnackbar={handleSnackbar} />
      </Provider>,
    );

    expect(screen.getByRole('dialog')).toBeInTheDocument();
    expect(screen.getByText('login.header')).toBeInTheDocument();
    expect(screen.getByText('login.email')).toBeInTheDocument();
    expect(screen.getByText('login.button')).toBeInTheDocument();
  });

  it('calls handleClose and handleSnackbar on successful login and updates the store', async () => {
    render(
      <Provider store={store}>
        <LoginModal isOpen={true} handleClose={handleClose} handleSnackbar={handleSnackbar} />
      </Provider>,
    );

    const emailInput = screen.getByRole('textbox', { name: /login\.email/i });

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });

    fireEvent.click(screen.getByText('login.button'));

    await waitFor(() => {
      expect(handleClose).toHaveBeenCalledWith(true);
      expect(handleSnackbar).toHaveBeenCalledWith(true);
      expect(sessionStorage.getItem('user')).toBe('test@example.com');
    });

    const state = store.getState();
    expect(state.user).toBe('test@example.com');
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <LoginModal isOpen={true} handleClose={handleClose} handleSnackbar={handleSnackbar} />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
