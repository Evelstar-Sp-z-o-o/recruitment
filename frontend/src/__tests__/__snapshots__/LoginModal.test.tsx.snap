// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginModal Component > should match snapshot 1`] = `<DocumentFragment />`;
